src/cudadecoder/decodable-cumatrix.h

```diff
@@ -0,0 +1,66 @@
+// cudadecoder/decodable-cumatrix.h
+/*
+ * Copyright (c) 2017, NVIDIA CORPORATION.  All rights reserved.
+ * Authors:  Hugo Braun, Justin Luitjens, Ryan Leary
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef KALDI_CUDA_DECODER_DECODABLE_CUMATRIX_H_
+#define KALDI_CUDA_DECODER_DECODABLE_CUMATRIX_H_
+
+#include "cudadecoder/cuda-decodable-itf.h"
+#include "cudamatrix/cu-matrix.h"
+#include "decoder/decodable-matrix.h"
+
+namespace kaldi {
+/**
+  Cuda Decodable matrix.  Takes transition model and posteriors and provides
+  an interface similar to the Decodable Interface
+  */
+class DecodableCuMatrixMapped : public CudaDecodableInterface {
+public:
+  // This constructor creates an object that will not delete "likes" when done.
+  // the frame_offset is the frame the row 0 of 'likes' corresponds to, would be
+  // greater than one if this is not the first chunk of likelihoods.
+  DecodableCuMatrixMapped(const TransitionModel &tm,
+                          const CuMatrixBase<BaseFloat> &likes,
+                          int32 frame_offset = 0);
+
+  virtual int32 NumFramesReady() const;
+
+  virtual bool IsLastFrame(int32 frame) const;
+
+  virtual BaseFloat LogLikelihood(int32 frame, int32 tid) {
+    KALDI_ASSERT(false);
```

Comment:
IMO because gcc is able to figure out (from our annotations) that KALDI_ASSERT does not return. But I might be mistaken.

Reply:
Since it is declared as non-void the standard says you need a return.  I don't think the language says that you don't need one if it never returns.  Like if you have an infinite loop, it's also required.  gcc/g++ is usually pretty good on this, but with centos, the gcc version is 10 years old or more.

Reply:
Yes gcc only triggers a warning when the return is missing, always found it weird. I’ll fix it, thanks for the info